src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs

```diff
@@ -81,31 +65,34 @@ private void VisitArrayType(IArrayTypeSymbol symbol, TypeWithAnnotations typeOpt
             }
         }
 
-        private void AddNullableAnnotations(TypeWithAnnotations typeOpt)
+        private void AddNullableAnnotations(ITypeSymbol type)
         {
-            if (ShouldAddNullableAnnotation(typeOpt))
+            if (ShouldAddNullableAnnotation(type))
             {
-                AddPunctuation(typeOpt.NullableAnnotation.IsAnnotated() ? SyntaxKind.QuestionToken : SyntaxKind.ExclamationToken);
+                AddPunctuation(type.NullableAnnotation == CodeAnalysis.NullableAnnotation.Annotated ? SyntaxKind.QuestionToken : SyntaxKind.ExclamationToken);
             }
         }
 
-        private bool ShouldAddNullableAnnotation(TypeWithAnnotations typeOpt)
+        private bool ShouldAddNullableAnnotation(ITypeSymbol type)
         {
-            if (!typeOpt.HasType)
-            {
-                return false;
-            }
-            else if (format.MiscellaneousOptions.IncludesOption(SymbolDisplayMiscellaneousOptions.IncludeNullableReferenceTypeModifier) &&
-                !typeOpt.IsNullableType() && !typeOpt.Type.IsValueType &&
-                typeOpt.NullableAnnotation.IsAnnotated())
+            switch (type.NullableAnnotation)
             {
-                return true;
-            }
-            else if (format.CompilerInternalOptions.IncludesOption(SymbolDisplayCompilerInternalOptions.IncludeNonNullableTypeModifier) &&
-                !typeOpt.Type.IsValueType &&
-                typeOpt.NullableAnnotation.IsNotAnnotated() && !typeOpt.Type.IsTypeParameterDisallowingAnnotation())
-            {
-                return true;
+                case CodeAnalysis.NullableAnnotation.Annotated:
+                    if (format.MiscellaneousOptions.IncludesOption(SymbolDisplayMiscellaneousOptions.IncludeNullableReferenceTypeModifier) &&
+                        !ITypeSymbolHelpers.IsNullableType(type) && !type.IsValueType)
```

Comment:
>IsNullableType [](start = 44, length = 14)

Possibly outside the scope of this PR, but can we rename this to `IsNullableValueType`? This is now quite confusing.

Reply:
> Possibly outside the scope of this PR, but can we rename this to IsNullableValueType? This is now quite confusing.

This is out of scope. We use this name since beginning of Roslyn.

---
In reply to: [339212556](https://github.com/dotnet/roslyn/pull/39498#discussion_r339212556) [](ancestors = 339212556)